hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala

```diff
@@ -402,6 +389,40 @@ object HoodieSparkSqlWriter {
     }
   }
 
+  def deduceOperation(hoodieConfig: HoodieConfig, paramsWithoutDefaults : Map[String, String]): WriteOperationType = {
+    var operation = WriteOperationType.fromValue(hoodieConfig.getString(OPERATION))
+    // TODO clean up
+    // It does not make sense to allow upsert() operation if INSERT_DROP_DUPS is true
+    // Auto-correct the operation to "insert" if OPERATION is set to "upsert" wrongly
+    // or not set (in which case it will be set as "upsert" by parametersWithWriteDefaults()) .
+    if (hoodieConfig.getBoolean(INSERT_DROP_DUPS) &&
+      operation == WriteOperationType.UPSERT) {
+
+      log.warn(s"$UPSERT_OPERATION_OPT_VAL is not applicable " +
+        s"when $INSERT_DROP_DUPS is set to be true, " +
+        s"overriding the $OPERATION to be $INSERT_OPERATION_OPT_VAL")
+
+      operation = WriteOperationType.INSERT
+    }
+
+    // if no record key, no preCombine and no explicit partition path is set, we should treat it as append only workload
+    // and make bulk_insert as operation type.
+    if (!paramsWithoutDefaults.containsKey(DataSourceWriteOptions.RECORDKEY_FIELD.key())
+      &&  !paramsWithoutDefaults.containsKey(DataSourceWriteOptions.PARTITIONPATH_FIELD.key())
+      && !paramsWithoutDefaults.containsKey(DataSourceWriteOptions.PRECOMBINE_FIELD.key())
+      && !paramsWithoutDefaults.containsKey(OPERATION.key())) {
+      log.warn(s"Choosing BULK_INSERT as the operation type since auto record key generation is applicable")
+      operation = WriteOperationType.BULK_INSERT
+    }
+    // if no record key is set, will switch the default operation to INSERT (auto record key gen)
+    else if (!hoodieConfig.contains(DataSourceWriteOptions.RECORDKEY_FIELD.key())
+      && !paramsWithoutDefaults.containsKey(OPERATION.key())) {
+      log.warn(s"Choosing INSERT as the operation type since auto record key generation is applicable")
+      operation = WriteOperationType.INSERT
```

Comment:
so here both of the cases are for auto keygen: partition and precombine not set => bulkinsert, either partition or precombine is set => insert

not really sure about the rationale behind; bulkinsert vs insert is only about using small file handling or not, which should not be coupled with whether partition or precombine set by user?

Reply:
as I mentioned in the other comment, we are mostly trying to give better perf for users coming from parquet world. If a user is explicitly setting record key configs, then we can assume they about hudi's diff operation and if their use-case is immutable, they would explicitly set bulk_insert.